osu.Game/Graphics/ParticleExplosion.cs

```diff
@@ -0,0 +1,144 @@
+        /// Supports transform time offset chaining.
+        /// </summary>
+        public void Restart()
+        {
+            startTime = TransformStartTime;
+            this.FadeOutFromOne(duration);
+
+            parts.Clear();
+            for (int i = 0; i < particleCount; i++)
+                parts.Add(new ParticlePart(duration));
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+            Invalidate(Invalidation.DrawNode);
+        }
+
+        protected override DrawNode CreateDrawNode() => new ParticleExplosionDrawNode(this);
+
+        private class ParticleExplosionDrawNode : SpriteDrawNode
+        {
+            private readonly List<ParticlePart> parts = new List<ParticlePart>();
+
+            private ParticleExplosion source => (ParticleExplosion)Source;
+
+            private double startTime;
+            private double currentTime;
+            private Vector2 sourceSize;
+
+            public ParticleExplosionDrawNode(Sprite source)
+                : base(source)
+            {
+            }
+
+            public override void ApplyState()
+            {
+                base.ApplyState();
+
+                parts.Clear();
+                parts.AddRange(source.parts);
+
+                sourceSize = source.Size;
+                startTime = source.startTime;
+                currentTime = source.Time.Current;
+            }
+
+            protected override void Blit(Action<TexturedVertex2D> vertexAction)
+            {
+                var time = currentTime - startTime;
+
+                foreach (var p in parts)
+                {
+                    Vector2 pos = p.PositionAtTime(time);
+                    float alpha = p.AlphaAtTime(time);
+
+                    var rect = new RectangleF(
+                        pos.X * sourceSize.X - Texture.DisplayWidth / 2,
+                        pos.Y * sourceSize.Y - Texture.DisplayHeight / 2,
+                        Texture.DisplayWidth,
+                        Texture.DisplayHeight);
+
+                    // convert to screen space.
+                    var quad = new Quad(
+                        Vector2Extensions.Transform(rect.TopLeft, DrawInfo.Matrix),
+                        Vector2Extensions.Transform(rect.TopRight, DrawInfo.Matrix),
+                        Vector2Extensions.Transform(rect.BottomLeft, DrawInfo.Matrix),
+                        Vector2Extensions.Transform(rect.BottomRight, DrawInfo.Matrix)
+                    );
+
+                    DrawQuad(Texture, quad, DrawColourInfo.Colour.MultiplyAlpha(alpha), null, vertexAction,
+                        new Vector2(InflationAmount.X / DrawRectangle.Width, InflationAmount.Y / DrawRectangle.Height),
+                        null, TextureCoords);
+                }
+            }
+        }
+
+        private readonly struct ParticlePart
+        {
+            private readonly double duration;
+            private readonly double direction;
```

Comment:
This can be changed to a `float` if I didn't miss something
 Then line 138 can be changed to
`
            return new Vector2(0.5f) + travelledDistance * new Vector2(MathF.Sin(direction), MathF.Cos(direction));
`